Refs #27064 -- Added RenameIndex migration operation.

David-Wobrock opened the pull request:
Adds for https://code.djangoproject.com/ticket/27064 the `RenameIndex` operation

Split from https://github.com/django/django/pull/15651. Here we implement the operation, in https://github.com/django/django/pull/15651 we use it in the autodetector.
This PR:

- Added the `RenameIndex` operation, that nearly always has 2 behaviours for each method: 1/ renaming with an `old_name` or renaming an unnamed index with `old_fields`
- With its state forward, that either modifies the `indexes` or both `indexes` + `index_together`
- And the database forward/backward operations: which will either try to introspect the indexes to fetch the name, or simply fetch the Index
- Added related unit tests
- Added the  `supports_index_renaming` feature flag on DB backends, to know if one can use a `RENAME INDEX` database statement or needs to drop/create the index
- Adapted the documentation

felixxm commented on django/db/migrations/operations/models.py:
```diff
@@ -876,6 +877,126 @@ def migration_name_fragment(self):
+    def migration_name_fragment(self):
+        if self.old_name:
+            return "rename_%s_%s" % (self.old_name_lower, self.new_name_lower)
+        return "rename_%s_%s" % ("_".join(self.old_fields), self.new_name_lower)
```
Should we add `reduce()` for `RenameIndex` on the same model with `self.new_name_lower == operation.old_name_lower`? :thinking:

David-Wobrock replied:
Not sure it will be used often, but let's add it for the sake of completeness :+1:

felixxm commented on django/db/migrations/state.py:
```diff
@@ -187,6 +187,15 @@ def alter_model_options(self, app_label, model_name, options, option_keys=None):
+            model_state.options[option_name] = [
+                obj for obj in objs if tuple(obj) != tuple(value_to_remove)
+            ]
+        self.reload_model(app_label, model_name, delay=True)
```
Can we use `alter_model_options()` with `{"index_together": None}` instead?

David-Wobrock replied:
I think we could use something like:
```
    def remove_model_options(self, app_label, model_name, option_name, value_to_remove):
        model_state = self.models[app_label, model_name]
        if option_name in model_state.options:
            objs = model_state.options[option_name]
            self.alter_model_options(app_label, model_name, {option_name: objs}
```

But I don't think we can entirely drop the `remove_model_options` for `{"index_together": None}`.
We don't necessarily remove _all_ `index_together` values. In theory, we could just rename one of many.
For instance if we go from a model looking like
```
class MyModel(models.Model):
[...]
    class Meta:
        index_together = (
            ("foo", bar"),
            ("bar", baz"),
        )
```
to
```
class MyModel(models.Model):
[...]
    class Meta:
        index_together = (
            ("foo", bar"),
        )
        indexes = [
            models.Index(fields=("bar", baz")),
        ]
```

felixxm commented on docs/ref/migration-operations.txt:
```diff
@@ -222,6 +222,21 @@ Creates an index in the database table for the model with ``model_name``.
+
+.. versionadded:: 4.1
+
+.. class:: RenameIndex(model_name, new_name, old_name, old_fields)
```
```suggestion
.. class:: RenameIndex(model_name, new_name, old_name=None, old_fields=None)
```

felixxm commented on tests/migrations/test_operations.py:
```diff
@@ -2900,6 +2900,99 @@ def test_remove_index(self):
+        """
+        Test the RenameIndex operation.
+        """
+        app_label = "test_rnin"
```
This docstring is unnecessary.

felixxm commented on tests/migrations/test_operations.py:
```diff
@@ -2900,6 +2900,99 @@ def test_remove_index(self):
+        )
+
+    def test_rename_index_unnamed_index(self):
+        """
+        Test the RenameIndex operation with an unnamed index.
+        """
```
This docstring is unnecessary.

felixxm commented on django/db/migrations/operations/models.py:
```diff
@@ -876,6 +877,126 @@ def migration_name_fragment(self):
+            matching_index_name = schema_editor._constraint_names(
+                from_model, column_names=self.old_fields, index=True
+            )
+            if len(matching_index_name) > 1:
+                raise AmbiguityError(
+                    f"Found more than one index on table {from_model._meta.db_table} "
+                    f"for columns {self.old_fields}."
+                )
+            if len(matching_index_name) == 0:
+                raise KeyError(
+                    f"Found no index on table {from_model._meta.db_table} "
+                    f"for columns {self.old_fields}."
+                )
```
Also these error messages are untested.

felixxm commented on django/db/migrations/operations/models.py:
```diff
@@ -876,6 +877,126 @@ def migration_name_fragment(self):
+    """Rename an index."""
+
+    def __init__(self, model_name, new_name, old_name=None, old_fields=None):
+        if not bool(old_name) ^ bool(old_fields):
+            raise ValueError(
+                "RenameIndex requires exactly one of old_name and old_fields arguments "
+                "to be set."
+            )
```
Can we also added tests for these error messages?

felixxm reviewed (commented):
@David-Wobrock Thanks :+1: I left comments.

David-Wobrock commented:
Thanks a lot for the review @felixxm :pray: 
I amended the commit, where I:
- removed docstrings and added the default arguments for the `RenameIndex` docs
- added tests for error messages and changed them to `ValueError`s
- added a `reduce` function and a unit test for it

felixxm reviewed (commented):
@David-Wobrock Thanks for updates :+1: I rebased and pushed some edits.

felixxm commented:
buildbot, test on oracle.

David-Wobrock commented:
Awesome, thanks a lot for the edits @felixxm :rocket:

David-Wobrock commented on docs/releases/4.1.txt:
```diff
@@ -337,7 +337,10 @@ Management Commands
+* The new :class:`~django.db.migrations.operations.RenameIndex` operation
+  allows renaming indexes defined in the :attr:`Meta.indexes
+  <django.db.models.Options.indexes>` or
+  :attr:`~django.db.models.Options.unique_together` options.
```
```suggestion
  :attr:`~django.db.models.Options.index_together` options.
```

felixxm replied:
Good catch :dart:
